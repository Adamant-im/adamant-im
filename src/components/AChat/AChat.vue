<template>
  <div class="a-chat">
    <div class="a-chat__content">

      <slot name="header"></slot>

      <v-divider/>

      <div class="a-chat__body">
        <div class="text-xs-center py-2">
          <v-progress-circular
            v-show="loading"
            indeterminate
            color="primary"
            size="24"
            style="z-index: 100"
          />
        </div>

        <div ref="messages" class="a-chat__body-messages">
          <slot name="messages" :messages="messages">

              <template v-for="message in messages">

                <slot name="message"
                  :message="message"
                  :sender="getSenderMeta(message.senderId)"
                  :user-id="userId"
                  :locale="locale"
                />

              </template>

          </slot>
        </div>

      </div>

      <slot name="form"></slot>
    </div>
  </div>
</template>

<script>
import AChatMessage from './AChatMessage'
import AChatTransaction from './AChatTransaction'

export default {
  mounted () {
    this.attachScrollListener()
  },
  beforeDestroy () {
    this.destroyScrollListener()
  },
  data: () => ({
    currentScrollHeight: 0,
    currentScrollTop: 0
  }),
  methods: {
    attachScrollListener () {
      this.$refs.messages.addEventListener('scroll', this.onScroll)
    },

    destroyScrollListener () {
      this.$refs.messages.removeEventListener('scroll', this.onScroll)
    },

    onScroll () {
      const scrollHeight = this.$refs.messages.scrollHeight
      const scrollTop = this.$refs.messages.scrollTop
      const clientHeight = this.$refs.messages.clientHeight

      // Scrolled to Bottom
      if (scrollHeight - scrollTop === clientHeight) {
        this.$emit('scroll:bottom')
      } else if (scrollTop === 0) { // Scrolled to Top
        // Save current `scrollHeight` to maintain scroll
        // position when unshift new messages
        this.currentScrollHeight = scrollHeight
        this.$emit('scroll:top')
      }

      // Save currentScrollTop.
      // Used when scrolled bottom while loading messages.
      this.currentScrollTop = this.$refs.messages.scrollTop
    },

    // Fix scroll position after unshift new messages.
    // Called from parent component.
    maintainScrollPosition () {
      this.$refs.messages.scrollTop =
        this.$refs.messages.scrollHeight - this.currentScrollHeight + this.currentScrollTop
    },

    // Scroll to Bottom when new message.
    // Called from parent component.
    scrollToBottom () {
      this.$refs.messages.scrollTop = this.$refs.messages.scrollHeight
    },

    /**
     * Returns sender address and name.
     * @param {string} senderId Sender address
     * @returns {{ id: string, name: string }}
     */
    getSenderMeta (senderId) {
      return this.partners.find(partner => partner.id === senderId)
    }
  },
  components: {
    AChatMessage,
    AChatTransaction
  },
  props: {
    messages: {
      type: Array,
      default: () => []
    },
    partners: {
      type: Array,
      default: () => []
    },
    userId: {
      type: String
    },
    loading: {
      type: Boolean,
      default: false
    },
    locale: {
      type: String,
      default: 'en'
    }
  }
}
</script>
