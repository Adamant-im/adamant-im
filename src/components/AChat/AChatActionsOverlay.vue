<template>
  <div :class="classes.root" ref="rootRef" @click="handleClick">
    <div
      :class="classes.message"
      :style="{
        top: position.top,
        left: position.left,
        width: position.width
      }"
    >
      <div
        :class="{
          [classes.reactionSelect]: true,
          [classes.reactionSelectLeft]: transaction.senderId === partnerId,
          [classes.reactionSelectBottom]: isLargeMessage
        }"
      >
        <slot name="top" />
      </div>

      <slot />

      <div
        :class="{
          [classes.menu]: true,
          [classes.menuLeft]: transaction.senderId === partnerId,
          [classes.menuBottom]: isLargeMessage
        }"
      >
        <slot name="bottom" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { usePartnerId } from '@/components/AChat/hooks/usePartnerId'
import { NormalizedChatMessageTransaction } from '@/lib/chat/helpers'
import { computed, defineComponent, onMounted, PropType, reactive, ref } from 'vue'

const className = 'a-chat-actions-overlay'
const classes = {
  root: className,
  message: `${className}__message`,
  reactionSelect: `${className}__reaction-select`,
  reactionSelectLeft: `${className}__reaction-select--left`,
  reactionSelectBottom: `${className}__reaction-select--bottom`,
  menu: `${className}__menu`,
  menuLeft: `${className}__menu--left`,
  menuBottom: `${className}__menu--bottom`
}

export default defineComponent({
  props: {
    modelValue: {
      type: Boolean,
      required: true
    },
    transaction: {
      type: Object as PropType<NormalizedChatMessageTransaction>,
      required: true
    }
  },
  emits: ['update:modelValue'],
  setup(props, { emit }) {
    const rootRef = ref<HTMLElement | null>(null)
    const partnerId = usePartnerId(props.transaction)

    const rect = computed(() => {
      const element = document.querySelector(`.a-chat__message[data-id='${props.transaction.id}']`)

      if (element) {
        return element.getBoundingClientRect()
      } else {
        console.warn(
          `[AChatActionsOverlay]: <AChatMessage/> with "data-id"="${props.transaction.id}" was not found`
        )
        return null
      }
    })

    const position = reactive({
      top: `${rect.value?.top}px`,
      left: `${rect.value?.left}px`,
      bottom: `${rect.value?.bottom}px`,
      right: `${rect.value?.right}px`,
      width: `${rect.value?.width}px`
    })

    const isLargeMessage = computed(() => {
      if (!rect.value) return false

      return rect.value.height > window.innerHeight / 2
    })

    onMounted(() => {
      setTimeout(() => {
        if (rect.value) {
          const halfOfElementHeight = rect.value ? rect.value.height : 0

          position.top = `calc(50% - ${halfOfElementHeight}px)`
        }
      }, 0)
    })

    const handleClick = (e: MouseEvent) => {
      e.stopPropagation()

      // if clicked outside the <AChatMessage/>
      // close overlay
      if (e.target === rootRef.value) {
        emit('update:modelValue', false)

        return
      }
    }

    return {
      classes,
      position,
      rootRef,
      handleClick,
      partnerId,
      isLargeMessage
    }
  }
})
</script>

<style lang="scss">
.a-chat-actions-overlay {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;

  &__message {
    position: fixed;
    transition: all 0.05s ease;
  }

  &__reaction-select {
    position: absolute;
    bottom: 100%;
    margin-bottom: 16px;
    z-index: 1;

    right: 0;
    left: unset;

    &--left {
      left: 0;
      right: unset;
    }

    &--bottom {
      bottom: unset;
      top: 100%;
    }
  }

  &__menu {
    position: absolute;
    top: 100%;

    right: 0;
    left: unset;

    &--left {
      left: 0;
      right: unset;
    }

    &--bottom {
      margin-top: calc(46px + 16px); // <AReactionSelect/> height + margin
    }
  }
}

.v-theme--light {
  .a-chat-actions-overlay {
  }
}

.v-theme--dark {
  .a-chat-actions-overlay {
  }
}
</style>
