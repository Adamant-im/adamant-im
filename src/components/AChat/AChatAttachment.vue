<template>
  <v-row
    class="a-chat__message-container"
    :class="{
      'a-chat__message-container--right': isStringEqualCI(transaction.senderId, userId),
      'a-chat__message-container--transition': elementLeftOffset === 0,
      'a-chat__message-container--disable-max-width': disableMaxWidth
    }"
    v-touch="{
      move: onMove,
      end: onSwipeEnd
    }"
    :style="{
      left: swipeDisabled ? '0px' : `${elementLeftOffset}px`
    }"
    v-longpress="onLongPress"
  >
    <div
      class="a-chat__message"
      :class="{
        'a-chat__message--flashing': flashing,
        'elevation-9': elevation
      }"
      :data-id="dataId"
    >
      <div
        v-if="showAvatar"
        class="a-chat__message-avatar hidden-xs-only"
        :class="{ 'a-chat__message-avatar--right': isStringEqualCI(transaction.senderId, userId) }"
      >
        <slot name="avatar" />
      </div>
      <div class="a-chat__message-card">
        <div class="a-chat__message-card-header mt-1">
          <div v-if="status.status === 'CONFIRMED'" class="a-chat__blockchain-status">&#x26AD;</div>
          <div class="a-chat__timestamp">
            {{ time }}
          </div>
          <div v-if="isOutgoingMessage" class="a-chat__status">
            <v-icon
              v-if="status.status === 'REJECTED'"
              :icon="statusIcon"
              :title="$t('chats.retry_message')"
              size="15"
              color="red"
              @click="$emit('resend')"
            />
            <v-icon v-else :icon="statusIcon" size="13" />
          </div>
        </div>

        <div v-if="transaction.isReply" class="a-chat__quoted-message">
          <quoted-message
            :message-id="transaction.asset.replyto_id"
            @click="$emit('click:quotedMessage', transaction.asset.replyto_id)"
          />
        </div>

        <div class="a-chat__message-card-body">
          <!-- eslint-disable vue/no-v-html -- Safe with DOMPurify.sanitize() content -->
          <!-- AChatMessage :message <- Chat.vue :message="formatMessage(message)" <- formatMessage <- DOMPurify.sanitize() -->
          <div
            v-if="html"
            class="a-chat__message-text a-text-regular-enlarged"
            v-html="formattedMessage"
          />
          <!-- eslint-enable vue/no-v-html -->
          <div
            v-else
            class="a-chat__message-text a-text-regular-enlarged"
            v-text="formattedMessage"
          />
          <v-btn @click="getFileFromStorage()">CLICK ME</v-btn>
          <pre>
            {{ JSON.stringify(transaction.id, null, 2) }}
          </pre>
        </div>
      </div>
    </div>

    <slot name="actions" />
  </v-row>
</template>

<script lang="ts">
import { computed, defineComponent, PropType } from 'vue'
import { useStore } from 'vuex'

import { useFormatMessage } from './hooks/useFormatMessage'
import { usePartnerId } from './hooks/usePartnerId'
import { useTransactionTime } from './hooks/useTransactionTime'
import { NormalizedChatMessageTransaction } from '@/lib/chat/helpers'
import { downloadFile, isStringEqualCI } from '@/lib/textHelpers'
import { tsIcon } from '@/lib/constants'
import QuotedMessage from './QuotedMessage.vue'
import { useSwipeLeft } from '@/hooks/useSwipeLeft'
import formatDate from '@/filters/date'
import { isWelcomeChat } from '@/lib/chat/meta/utils'

export default defineComponent({
  methods: { downloadFile },
  components: {
    QuotedMessage
  },
  props: {
    transaction: {
      type: Object as PropType<NormalizedChatMessageTransaction>,
      required: true
    },
    dataId: {
      type: String
    },
    status: {
      type: Object,
      required: true
    },
    html: {
      type: Boolean,
      default: false
    },
    /**
     * Highlight the message by applying a background flash effect
     */
    flashing: {
      type: Boolean,
      default: false
    },
    disableMaxWidth: {
      type: Boolean
    },
    elevation: {
      type: Boolean
    },
    swipeDisabled: {
      type: Boolean
    }
  },
  emits: ['resend', 'click:quotedMessage', 'swipe:left', 'longpress'],
  setup(props, { emit }) {
    const store = useStore()

    const userId = computed(() => store.state.address)
    const partnerId = usePartnerId(props.transaction)

    const getFileFromStorage = async () => {
      //TODO: refactor for each file
      const cid = props.transaction.asset.files[0].file_id
      const fileName = props.transaction.asset.files[0].file_name
      const fileType = props.transaction.asset.files[0].file_type
      const nonce = props.transaction.asset.files[0].nonce
      const myAddress = store.state.address

      const publicKey =
        props.transaction.senderId === myAddress
          ? props.transaction.recipientPublicKey
          : props.transaction.senderPublicKey
      const data = await store.dispatch('attachment/getAttachment', { cid, publicKey, nonce })
      if (!!data && !!fileName && !!fileType) {
        // TODO: resolve MIME-type
        downloadFile(data, fileName, '')
      }
    }

    const showAvatar = computed(() => !isWelcomeChat(partnerId.value))

    const statusIcon = computed(() => tsIcon(props.status.virtualStatus))
    const isOutgoingMessage = computed(() =>
      isStringEqualCI(props.transaction.senderId, userId.value)
    )
    const formattedMessage = useFormatMessage(props.transaction)
    const time = useTransactionTime(props.transaction)

    const { onMove, onSwipeEnd, elementLeftOffset } = useSwipeLeft(() => {
      emit('swipe:left')
    })

    const onLongPress = () => {
      emit('longpress')
    }

    return {
      userId,
      statusIcon,
      isOutgoingMessage,
      formattedMessage,
      getFileFromStorage,
      showAvatar,
      onMove,
      onSwipeEnd,
      elementLeftOffset,
      isStringEqualCI,
      onLongPress,
      formatDate,
      time
    }
  }
})
</script>
