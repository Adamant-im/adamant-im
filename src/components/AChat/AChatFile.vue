<template>
  <div :class="classes.container">
    <AChatFileLoader :partner-id="partnerId" :file="img" :transaction="transaction">
      <template #default="{ isLoading, fileUrl, width, height }">
        <v-img :src="fileUrl" :width="400" :aspect-ratio="width / height">
          <template #placeholder>
            <div class="d-flex align-center justify-center fill-height">
              <v-progress-circular color="grey-lighten-4" indeterminate />
            </div>
          </template>
        </v-img>
      </template>
    </AChatFileLoader>
  </div>
</template>

<script lang="ts">
import { AChatFileLoader } from './AChatFileLoader.tsx'
import { defineComponent, ref, PropType, onMounted } from 'vue'
import { LocalFile, NormalizedChatMessageTransaction } from '@/lib/chat/helpers'
import { useStore } from 'vuex'
import { FileAsset } from '@/lib/adamant-api/asset'

function imgToDataURL(file: File): Promise<string> {
  return new Promise((resolve) => {
    const reader = new FileReader()

    reader.onload = (event) => {
      const dataURL = event.target?.result as string
      resolve(dataURL)
    }

    reader.readAsDataURL(file)
  })
}

function isLocalFile(file: FileAsset | LocalFile): file is LocalFile {
  return 'file' in file && file.file instanceof File
}

const className = 'a-chat-file'
const classes = {
  root: className,
  container: `${className}__container`,
  containerWithElement: `${className}__container-with-element`,
  img: `${className}__img`
}

export default defineComponent({
  props: {
    transaction: {
      type: Object as PropType<NormalizedChatMessageTransaction>,
      required: true
    },
    img: {
      type: Object as PropType<FileAsset | LocalFile>,
      required: true
    },
    partnerId: {
      type: String,
      required: true
    }
  },
  components: { AChatFileLoader },
  setup(props) {
    const store = useStore()
    const imageUrl = ref('')

    const getFileFromStorage = async () => {
      if (isLocalFile(props.img)) {
        imageUrl.value = await imgToDataURL(props.img.file.file)
        return
      }

      const myAddress = store.state.address

      const cid = props.img?.id
      const fileName = props.img?.name
      const fileType = props.img?.extension
      const nonce = props.img?.nonce

      const publicKey =
        props.transaction.senderId === myAddress
          ? props.transaction.recipientPublicKey
          : props.transaction.senderPublicKey
      imageUrl.value = await store.dispatch('attachment/getAttachmentUrl', {
        cid,
        publicKey,
        nonce
      })
      if (!!fileName && !!fileType) {
        // TODO: resolve MIME-type
        // downloadFile(data, fileName, '')
      }
    }

    onMounted(() => {
      getFileFromStorage()
    })

    return {
      classes,
      getFileFromStorage,
      imageUrl
    }
  }
})
</script>

<style lang="scss" scoped>
@import '@/assets/styles/settings/_colors.scss';
@import '@/assets/styles/themes/adamant/_mixins.scss';

.a-chat-file {
  border-left: 3px solid map-get($adm-colors, 'attention');
  border-radius: 8px;
  margin: 8px;

  &__container {
    max-width: 230px;
  }

  &__container-with-element {
    display: grid;
    gap: 2px;
    grid-template-columns: repeat(2, minmax(50px, 1fr));
    grid-template-rows: auto;
    max-height: 400px;
  }

  &__img {
    width: 100%;
    height: 100%;

    :deep(.v-img__img) {
      object-fit: fill;
    }
  }

  @if length(img) % 2 != 0 {
    &__img:nth-last-child(1):nth-child(odd) {
      grid-column: span 2;
    }
  }
}

.v-theme--light {
  .a-chat-file {
    background-color: map-get($adm-colors, 'secondary');
    color: map-get($adm-colors, 'regular');
  }
}

.v-theme--dark {
  .a-chat-file {
    background-color: rgba(245, 245, 245, 0.1); // @todo const
    color: #fff;
  }
}
</style>
