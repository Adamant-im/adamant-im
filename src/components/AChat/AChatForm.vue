<template>
  <div class="a-chat__form">
    <v-divider v-if="showDivider" class="a-chat__divider"/>
    <v-textarea
      v-model="message"
      v-on="listeners"
      :label="label"
      hide-details
      single-line
      auto-grow
      ref="messageTextarea"
      rows="1"

      :append-icon="showSendButton ? 'mdi-send' : ''"
      @click:append="submitMessage"
    >
      <template slot="prepend">
        <slot name="prepend"></slot>
      </template>
    </v-textarea>
  </div>
</template>

<script>
export default {
  computed: {
    /**
     * Processing `ctrl+enter`, `shift + enter` and `enter`
     */
    listeners () {
      return {
        keydown: (e) => {
          if (e.code === 'Enter') {
            if (this.sendOnEnter) {
              if (e.ctrlKey) { // add LF and calculate height when CTRL+ENTER
                this.addLineFeed()
                this.calculateInputHeight()
                return
              }

              if (!e.shiftKey) { // send message if shiftKey is not pressed
                e.preventDefault()
                this.submitMessage()
              }
            } else {
              if (e.ctrlKey || e.shiftKey) {
                e.preventDefault()
                this.submitMessage()
              }
            }
          }
        }
      }
    }
  },
  data: () => ({
    message: ''
  }),
  methods: {
    submitMessage () {
      this.$emit('message', this.message)
      this.message = ''
      // Fix textarea height to 1 row after miltiline message send
      this.calculateInputHeight()
    },
    calculateInputHeight () {
      this.$nextTick(this.$refs.messageTextarea.calculateInputHeight)
    },
    addLineFeed () {
      this.message += '\n'
    }
  },
  props: {
    showSendButton: {
      type: Boolean,
      default: true
    },
    sendOnEnter: {
      type: Boolean,
      default: true
    },
    label: {
      type: String,
      default: 'Type your message'
    },
    showDivider: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="stylus" scoped>
/**
 * Remove extra margins & paddings.
 */
.v-chat__form >>> .v-text-field
  padding-top: 0
  margin-top: 0

.a-chat__divider
  position: absolute
  top: 0
  left: 0
  width: 100%
</style>
