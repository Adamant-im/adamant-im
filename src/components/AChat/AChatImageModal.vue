<template>
  <div>
    <v-dialog :class="className" v-model="show">
      <div :class="`${className}__container`">
        <v-btn icon="mdi-arrow-collapse-down" :class="`${className}__btn-save`"></v-btn>
        <v-btn
          icon="mdi-window-close"
          :class="`${className}__btn-close`"
          @click="closeModal"
        ></v-btn>
        <p :class="`${className}__number-of-img`">{{ slide + 1 }} of {{ images.length }}</p>
        <v-carousel v-model="slide" :class="`${className}__carousel`" hide-delimiters>
          <v-carousel-item v-for="(item, i) in images" :key="i">
            <v-img :class="`${className}__modal-img`" :src="item.content" alt="Image" />
          </v-carousel-item>
        </v-carousel>
      </div>
    </v-dialog>
  </div>
</template>

<script>
export default {
  props: {
    images: Array,
    index: Number,
    modal: Boolean
  },
  emits: ['close', 'update:modal'],
  data: () => ({
    slide: null
  }),
  mounted() {
    this.slide = this.index
  },
  computed: {
    className: () => 'modal',
    show: {
      get() {
        return this.modal
      },
      set(value) {
        this.$emit('update:modal', value)
      }
    }
  },
  methods: {
    closeModal() {
      this.$emit('close')
    }
  }
}
</script>

<style lang="scss" scoped>
@import 'vuetify/settings';
.modal {
  z-index: 9999;
  max-width: 800px;

  &__container {
    position: relative;
    background-color: rgba(0, 0, 0, 0.54);
    width: 100%;
    margin: auto;
  }
  &__btn-close {
    position: absolute;
    right: 0;
    top: 0;
    margin: 4px;
    background-color: transparent;
  }
  &__btn-save {
    position: absolute;
    right: 0px;
    bottom: 0;
    margin: 4px;
    background-color: transparent;
  }
  &__carousel {
    margin-top: 40px;
    margin-bottom: 40px;
  }
  &__number-of-img {
    text-align: center;
    line-height: 50px;
    height: 50px;
  }
  &__modal-img {
    height: 100%;
    width: 100%;
  }
}
</style>
