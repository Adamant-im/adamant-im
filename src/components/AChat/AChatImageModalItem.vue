<template>
  <v-img class="img" :src="imageUrl" alt="Image" />
</template>
<script lang="ts">
import { defineComponent, PropType, ref, onMounted } from 'vue'
import { NormalizedChatMessageTransaction } from '@/lib/chat/helpers'
import { useStore } from 'vuex'
export default defineComponent({
  props: {
    file: {
      type: Object,
      required: true
    },
    transaction: {
      type: Object as PropType<NormalizedChatMessageTransaction>,
      required: true
    }
  },
  setup(props) {
    const store = useStore()
    const imageUrl = ref('')

    const getFileFromStorage = async () => {
      const myAddress = store.state.address

      const cid = props.file?.id
      const fileName = props.file?.name
      const fileType = props.file?.type
      const nonce = props.file?.nonce

      const publicKey =
        props.transaction.senderId === myAddress
          ? props.transaction.recipientPublicKey
          : props.transaction.senderPublicKey
      imageUrl.value = await store.dispatch('attachment/getAttachmentUrl', {
        cid,
        publicKey,
        nonce
      })
      if (!!fileName && !!fileType) {
        // TODO: resolve MIME-type
        // downloadFile(data, fileName, '')
      }
    }

    onMounted(() => {
      getFileFromStorage()
    })

    return {
      getFileFromStorage,
      imageUrl
    }
  }
})
</script>
<style scoped>
.img {
  height: 100%;
  width: 100%;
}
</style>
