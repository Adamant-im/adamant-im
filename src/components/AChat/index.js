import AChat from './AChat.vue'
import AChatForm from './AChatForm.vue'
import AChatMessage from './AChatMessage.vue'
import AChatTransaction from './AChatTransaction.vue'

export {
  AChat,
  AChatForm,
  AChatMessage,
  AChatTransaction
}
