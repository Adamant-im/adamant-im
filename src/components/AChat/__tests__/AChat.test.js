import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Vuetify from 'vuetify'

import AChat from '../AChat.vue'

const localVue = createLocalVue()
localVue.use(Vuetify)

describe('AChat.vue', () => {
  it('renders the correct markup', () => {
    const wrapper = shallowMount(AChat, {
      localVue
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should set ref to `messages`', () => {
    const wrapper = mount(AChat, {
      localVue
    })

    expect(wrapper.vm.$refs.messages).toBeTruthy()
  })

  // @todo More tests
  // @todo fix error `[Vuetify] Multiple instances of Vue detected`
})
