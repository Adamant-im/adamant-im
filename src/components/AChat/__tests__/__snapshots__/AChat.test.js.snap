// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AChat.vue > renders the correct markup 1`] = `
<div
  class="a-chat"
>
  <div
    class="a-chat__content"
  >
    
    
    <v-divider />
    <div
      class="a-chat__body"
    >
      <div
        class="text-center py-2"
      >
        <v-progress-circular
          color="primary"
          indeterminate=""
          size="24"
          style="z-index: 100; display: none;"
        />
      </div>
      <div
        class="a-chat__body-messages"
      >
        
        
        
        
      </div>
      <div
        class="a-chat__fab"
      >
        
        
      </div>
    </div>
    
    
  </div>
</div>
`;
